### ISSUE TITLE
Ability to map a DbFunction to a fragment of SQL

### ISSUE THREAD

**Issue by mokeyish** [dotnet/efcore#13337]
In this issue we discussed having the ability to map a DbFunction directly to an arbitrary fragment of SQL without requiring that the SQL is generated by the provider form expression nodes. We ended up deciding not to do this for various reasons:

1. It would be hard to adapt the current way SQL is generated to support this
2. As providers add more capabilities, the usefulness of this would diminish
3. With some databases it should already be possible to create user-defined function on the database that encapsulate the desired SQL fragment, and then map that function to a DbFunction

The latter approach is what we are recommending the customer to try. 

### Original issue as reported by customer
```SQL
select * from "SampleTable"
where !IsDelete and
"Permission"::jsonb -> 'roles' ?| array[
        '797da808-c739-4581-b0bf-12670fd4223a',
        '543e7b6b-f68f-4f24-92b0-463828d3de05']
```
 I want to add `"Permission"::jsonb -> 'roles' ?| array['797da808-c739-4581-b0bf-12670fd4223a','543e7b6b-f68f-4f24-92b0-463828d3de05']` to here as global filter. 
```C#
protected override void OnModelCreating(ModelBuilder modelBuilder) {
    modelBuilder.Entity<SampleTable>().HasQueryFilter(o => !o.IsDelete);
}
```

**Comment by ajcvickers** [comment]
@mokeyish This isn't possible with the current code, but putting it on the backlog to consider for the future.

**Comment by mokeyish** [comment]
@ajcvickers Is there an alternative solution to do this? Or how to change the final query sql.

**Comment by ajcvickers** [comment]
@smitpatel @maumar Is there any way to do this currently?

**Comment by smitpatel** [comment]
Query filters are not special constructs. You can have filter to do anything which can be expressed in a normal LINQ query. If you can write a linq query with filter without using global filter, I can provide a way to achieve it in global filter.

**Comment by mokeyish** [comment]
@smitpatel What's the way?

**Comment by smitpatel** [comment]
>  If you can write a linq query with filter without using global filter

@mokeyish - That is my question to you. How do you intend to write such filter in Where Clause in your linq query?

I removed this from backlog milestone. It is not clear what we are tracking here. Arbitrary SQL in Where predicate only?

**Comment by ajcvickers** [comment]
@smitpatel We're tracking this to specify a query filter using raw SQL. It's not specific to the actual example above, but the example above is what @mokeyish is trying to find a workaround for.

**Comment by ajcvickers** [comment]
@divega to follow up

**Comment by mokeyish** [comment]
Is it possible to register a named `function` and defined how to translated it to sql.Such as `DbFunctions.RawPredicateSql`

**Comment by ajcvickers** [comment]
@roji Question from triage: can the SQL above be encapsulated in a DbFunction?

**Comment by roji** [comment]
@ajcvickers yes, that should be a good workaround... IIRC the about-to-be-released 2.2 passes all DbFunction-related functional tests.

**Comment by roji** [comment]
@divega, I'm not sure exactly to which part of the query you're referring to, but yes - PostgreSQL should support parameters in the array (did a quick un-conclusive check).

In PostgreSQL, this query:
```
"Permission"::jsonb -> 'roles' ?| array[
        '797da808-c739-4581-b0bf-12670fd4223a',
        '543e7b6b-f68f-4f24-92b0-463828d3de05']
```

Means go to the `Permissions` column, look up the `roles` node at the root, and return true if that node contains one of the two GUIDs in the given array (see [the docs for the ?| operator](https://www.postgresql.org/docs/current/functions-json.html)). There's no reason why you shouldn't be able to write the following (`$1` is the native PostgreSQL parameter placeholder):

```sql
SELECT 1 WHERE '{"a":1, "b":2, "c":3}'::jsonb ?| array['a', $1];
```

Note that the Npgsql provider isn't yet anywhere near being able to translate anything into this - but if you wrap this SQL in a function and use that, then the function could accept a parameter indeed.

(I'm assuming you're no longer talking about global filters anymore, because those don't have parameters, right?)

**Comment by divega** [comment]
Thanks @roji! I believe you answered what I asked, but I am still unsure how to map the DbFunction :smile: 

TL;DR: I talked to @smitpatel about this, and he suggested that the simplest workaround would then be for the customer to embed this logic:

``` SQL
s."Permission"::jsonb -> 'roles' ?| array[
        '797da808-c739-4581-b0bf-12670fd4223a',
        '543e7b6b-f68f-4f24-92b0-463828d3de05']
```

into a PostgreSQL user-defined function with parameters. Assuming this is possible, then you can simply map calls to that function to a method, e.g. using the `[DbFunction]` attribute. 

I still suspect it could be useful to allow DbFunctions to map to in-lined SQL fragments that don't follow the shape of functions (i.e. that don't look like _function(arg1, arg2, ...)_), but according to @smitpatel it would be somewhat difficult to build something that is general purpose, because SQL generation would need to be able to incrementally emit the arguments and the substrings that go between the arguments. 

Anyway, this is going back to triage to decide whether we want to keep it in the backlog. 

cc @pmiddleton

**Comment by divega** [comment]
@roji re your question about query filters, in general, if you can pass a variable in a predicate in a LINQ query, you can pass a member of the derived DbContext in the equivalent query filter, and it will become a parameter.

**Comment by roji** [comment]
> @roji re your question about query filters, in general, if you can pass a variable in a predicate in a LINQ query, you can pass a member of the derived DbContext in the equivalent query filter, and it will become a parameter.

Oh, I didn't realize that... great.

Just to come back to the original question here, the OP doesn't necessarily seem to be interested in passing parameters - they seem to want to simply embed raw, unparameterized SQL in a global filter. It's good to have the DbFunction workaround (with or without parameterization).

**Comment by pmiddleton** [comment]
@divega - I can see how it would be useful to be able to emit raw sql into a where clause.  

However as @smitpatel said it is going to be very problematic to implement.  Beyond what he said there will also be no way for the end user to know what the surrounding emitted sql is going to look like.  Therefore they won't know things like what table aliases are being used, if clauses were lifted or pushed down, etc.  That is going to make writing a valid fragment tough, esp since there is no guarantee those parts of the query won't change from version to version of EF.

Thankfully the user doesn't need to pass in the guids they want to check because currently DbFunctions don't allow passing arrays of parameters.  This is something we can look at adding, but will require investigation of if/how the various databases support this.  With Sql Server you can do it but it requires defining a custom Type and then letting the SqlParameter know it is dealing with a custom type.  I'm not sure about other databases.

**Comment by divega** [comment]
> Therefore they won't know things like what table aliases are being used

That makes me wonder if we are handling things correctly for the free text functions we support. The first argument in the LINQ query is always a property reference, but in the SQL it has to translate to the right table and potentially column aliases.

> Thankfully the user doesn't need to pass in the guids they want to check because currently DbFunctions don't allow passing arrays of parameters

FWIW, @smitpatel and I came up with the idea of having a fixed number of discrete guid arguments in the function. It seems reasonable as long as you don’t need too many. Anyway, I think he implied that with the new approach in 3.0, the level of difficulty of mapping functions like this could be different.

**Comment by roji** [comment]
@pmiddleton

> Thankfully the user doesn't need to pass in the guids they want to check because currently DbFunctions don't allow passing arrays of parameters. This is something we can look at adding, but will require investigation of if/how the various databases support this. With Sql Server you can do it but it requires defining a custom Type and then letting the SqlParameter know it is dealing with a custom type. I'm not sure about other databases.

At least on the PostgreSQL side, there's first class support for array values - for every database type `x` you have a corresponding `x[]`. This is supported both at the Npgsql ADO level and at the EF Core level, so passing a (variable-width!) array parameter doesn't differ from passing in a string parameter. I don't see any reason for DbFunction to not support this, but I don't know that part of the code very well.

As long as we're on the subject, we could also think about supporting user-defined functions with variable number of parameters (i.e. C# `param`, or what PostgreSQL calls "variadic" functions). But this is a different subject... I can open a new issue for this, is there some general issue tracking open questions for functions for 3.0

**Comment by ajcvickers** [comment]
@divega to update description.
